### ISSUE TITLE
Running out of space on image during preload

### ISSUE THREAD

**Issue by BrianAdams** [resin-io/resin-preload-image-script#54]
When I download the docker image to my desktop to check the size I see that is it *1.91GB*, yet when doing the resin-preload, the script calculates the size as *798MB*.  I am assuming that is why it is running out of space when downloading the image:

```
docker images
REPOSITORY                                                                     TAG                 IMAGE ID            CREATED             SIZE
registry2.resin.io/...project.../86a155bc9b33e6f1d9710063638e26dc22921f42   latest              0cc797f6ebff        2 days ago          1.91GB
```

```
resin os download raspberrypi3 -o resin.img --version default
docker run --rm -e API_KEY=..key.. -e APP_ID=..appid.. -v /tmp/resin.img:/img/resin.img --privileged resin/resin-preload
```

```
Fetching application data
Using API host https://api.resin.io
Using Registry host registry2.resin.io
Using API_KEY
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   840  100   840    0     0    874      0 --:--:-- --:--:-- --:--:--  1746
...
Container size: 798 MB
Mapping /img/resin.img 46137344:327155712 to /dev/loop0

Mounting rootfs from /dev/loop0 to /mnt/202699-rootfs

Sourcing release info from /mnt/202699-rootfs/etc/os-release
Detected Resin OS 2.0.6+rev3

id: resin-os
name: Resin OS
version: 2.0.6+rev3
version_id: 2.0.6
pretty_name: Resin OS 2.0.6+rev3
resin_board_rev: 5f3fbb3
meta_resin_rev: c653cc2
slug: raspberrypi3
machine: raspberrypi3


Unmounting rootfs from /mnt/202699-rootfs
umount: /mnt/202699-rootfs unmounted
Unmapping /dev/loop0
Model:  (file)
Disk /img/resin.img: 1804MB
Sector size (logical/physical): 512B/512B
Partition Table: msdos
Disk Flags:

Number  Start   End     Size    Type      File system  Flags
 1      4.19MB  46.1MB  41.9MB  primary   fat16        boot, lba
 2      46.1MB  373MB   327MB   primary   ext4
 3      373MB   700MB   327MB   primary
 4      700MB   1804MB  1103MB  extended               lba
 5      705MB   726MB   21.0MB  logical   ext4
 6      730MB   1804MB  1074MB  logical   ext4

Expanding image by 877 MB
877+0 records in
877+0 records out
Expanding extended partition 4 by 877 MB
Expanding logical partition 6 by 877 MB
Model:  (file)
Disk /img/resin.img: 2723MB
Sector size (logical/physical): 512B/512B
Partition Table: msdos
Disk Flags:

Number  Start   End     Size    Type      File system  Flags
 1      4.19MB  46.1MB  41.9MB  primary   fat16        boot, lba
 2      46.1MB  373MB   327MB   primary   ext4
 3      373MB   700MB   327MB   primary
 4      700MB   2681MB  1981MB  extended               lba
 5      705MB   726MB   21.0MB  logical   ext4
 6      730MB   2681MB  1951MB  logical   ext4

Creating mountpoint /mnt/202699
Expanding filesystem
Using EXTFS
Mapping /img/resin.img 729808896:1951191552 to /dev/loop0
Using /dev/loop0
Resizing filesystem
e2fsck 1.43 (17-May-2016)
e2fsck: need terminal for interactive repairs
Removing /mnt/202699/rce
Waiting for Docker to start...
time="2017-06-23T16:28:49.186529914Z" level=info msg="Graph migration to content-addressability took 0.00 seconds"
time="2017-06-23T16:28:49.189610610Z" level=warning msg="Running modprobe bridge br_netfilter failed with message: modprobe: can't change directory to '/lib/modules': No such file or directory\n, error: exit status 1"
time="2017-06-23T16:28:49.191319399Z" level=warning msg="Running modprobe nf_nat failed with message: `modprobe: can't change directory to '/lib/modules': No such file or directory`, error: exit status 1"
time="2017-06-23T16:28:49.259926136Z" level=info msg="Default bridge (docker0) is assigned with an IP address 172.18.0.0/16. Daemon option --bip can be used to set a preferred IP address"
time="2017-06-23T16:28:49.321050917Z" level=info msg="Loading containers: start."

time="2017-06-23T16:28:49.321929276Z" level=info msg="Loading containers: done."
time="2017-06-23T16:28:49.322046467Z" level=info msg="Daemon has completed initialization"
time="2017-06-23T16:28:49.322337814Z" level=info msg="Docker daemon" commit=20f81dd execdriver=native-0.2 graphdriver=aufs version=1.10.3
time="2017-06-23T16:28:49.338344333Z" level=info msg="API listen on /tmp/docker-202699/docker.sock"
Docker started
Pulling image...
...

a1f1e8c37f14: Verifying Checksum
a1f1e8c37f14: Download complete
time="2017-06-23T16:30:14.848362139Z" level=error msg="Download failed: write /mnt/202699/docker/tmp/GetImageBlob199650751: no space left on device"
a1f1e8c37f14: Download complete

```

**Comment by BrianAdams** [comment]
From a quick glance at the way the size of the image is being calculated in the script I wonder if the problem is that it is using content-length of the compressed payload. Just a hunch.
https://stackoverflow.com/questions/3819280/content-length-when-using-http-compression

**Comment by jhermsmeier** [comment]
@BrianAdams can you try the `fix-container-size` branch (from #55), and see if that does it?

**Comment by BrianAdams** [comment]
@jhermsmeier no luck with the fix-container-size branch. Still calculates the same value. Perhaps my hunch is off.

**Comment by BrianAdams** [comment]
I also noticed that Docker Hub reports that size 800MB size as the compressed size.

<img width="571" alt="openrov_rov-image_-_docker_hub" src="https://user-images.githubusercontent.com/1471826/27503468-29aa4aba-5831-11e7-9d49-1f9522c00a6b.png">

**Comment by BrianAdams** [comment]
Looks like the Docker V2 registries now store layers in tar.gz compressed formats now. This same issue hit Docker HUB when displaying image sizes: https://github.com/docker/hub-feedback/issues/331

**Comment by BrianAdams** [comment]
I went ahead and forced the image to expand the amount on the uncompressed image but it turns out that is not helping the download.  I still get the out of space issue:
```
time="2017-06-27T23:01:08.102031114Z" level=error msg="Download failed, retrying: write /mnt/202699/docker/tmp/GetImageBlob421434463: no space left on device" 
```
Unfortunately I've taken this as far as I can at this point. 'df' and all other commands in the image are unable to interrogate the AUFS partition, so I'm not sure how to validate the space usage.

The best I was able to do is to run du on the mount while it was running to see if died while only taking up  936MB of what should have been 3104MB free.
```
936.1M	/mnt/202699/docker
936.2M	/mnt/202699/
936.2M	total
```
```

Number  Start   End     Size    Type      File system  Flags
 1      4.19MB  46.1MB  41.9MB  primary   fat16        boot, lba
 2      46.1MB  373MB   327MB   primary   ext4
 3      373MB   700MB   327MB   primary
 4      700MB   3807MB  3107MB  extended               lba
 5      705MB   726MB   21.0MB  logical   ext4
 6      730MB   3807MB  3077MB  logical   ext4
```

**Comment by jhermsmeier** [comment]
@BrianAdams thanks for all the research & details on this – I've been trying to get around this the past two days, but from what I understand, there's no way of getting the uncompressed layer sizes through the docker registry v2, apart from downloading & decompressing them all (which doesn't feel like a viable alternative to me either).

**Comment by jhermsmeier** [comment]
The tarballs would theoretically allow us to request the last 4 bytes of a layer (gzip has a uint32 at the end which is it's uncompressed size) via the ranges-header, and then calculate it from that. There's one very significant drawback with that as well though, which is that gzip can't represent uncompressed sizes of 4GB and more – meaning a layer's size, if equal to or above 4 GB, would be calculated as `uncompressedSize % (2^32)` (in the worst case zero bytes), resulting in the same issue.

**Comment by jhermsmeier** [comment]
Yet another workaround could be to multiply the compressed size by the average gzip compression ratio, plus a margin – but that's also rather dangerous, albeit prob. less than what currently happens.

**Comment by BrianAdams** [comment]
@jhermsmeier Thanks for the update!  Please also note that there are potentially two different issues here. 

1. Are we reserving enough space for the docker container image, which the investigation around how to determine the on-disk space is relevant.  I went ahead and hard-coded a value in the script that should have worked around this issue, which exposed issue 2.

2. Even when the img file is expanded enough to hold the actual on disk size of the of expanded docker image, the download of that image is still failing with an out of disk space issue.

I can work around issue 1, I've added the CONTAINER_SIZE as an environment variable that I can pass in that overrides the calculating of the container size.  Issue 2 is blocking me right now, as have not found a work around to the download failing with an out of space error message when there is presumably plenty of space left.

**Comment by jhermsmeier** [comment]
@BrianAdams I just spotted something in the log you posted;

```
Resizing filesystem
e2fsck 1.43 (17-May-2016)
e2fsck: need terminal for interactive repairs
Removing /mnt/202699/rce
```

This means the partition's filesystem wasn't expanded, even though the partition has been resized – hence it runs out of space.
Can you check / tell me if that's the case for the other preload attempts with your image as well?
I'll try and see if I can reproduce with an image of that resinOS version.

I'm also wondering why the mountpoint is called `/mnt/202699/rce` and not `/mnt/202699/docker` – I think that shouldn't happen if it's resinOS 2.x (which it looks like from the log). Looking into that as well now.

**Comment by BrianAdams** [comment]
So it turns out that the e2fsck warning is happening on both the linux and osx based build environments I have tried this on. 

Could this be an issue with using AUFS partitions on the docker-1.10 base Alpine image that does not have aufs tools installed?

**Comment by BrianAdams** [comment]
@jhermsmeier and just confirming, yes these are resinOS 2.x images.

**Comment by jhermsmeier** [comment]
> So it turns out that the e2fsck warning is happening on both the linux and osx based build environments I have tried this on.

That'll be an issue with the resinOS image then – maybe it got corrupted during download / transfer, or it's something further up the pipeline on our end. I can't reproduce it though, fs expansion works fine for me – just downloaded a fresh image from the dashboard and it all works.

> Could this be an issue with using AUFS partitions on the docker-1.10 base Alpine image that does not have aufs tools installed?

No, the AUFS storage driver is part of docker, and it's just a union-filesystem on top of other file systems (in this case ext4), and the preloader uses the e2fs progs to do the filesystem resizing of the ext4 data partitions.

So yeah..., most likely it's the image that's being preloaded that is corrupted.
Does the same happen for you with a fresh downloaded image?

**Comment by jhermsmeier** [comment]
Just an fyi: Fixed the `rce` removal issue I spotted in https://github.com/resin-io/resin-preload-image-script/pull/56 – has no impact in this case though.

**Comment by BrianAdams** [comment]
So I downloaded the resin os image yet again:
```
resin os download raspberrypi3 -o resin.img --version menu
Getting device operating system for raspberrypi3
? Select the OS version: v2.0.6+rev3.prod (recommended)
The image was downloaded successfully
```

Still fails to expand the storage.  Working on getting v2.0.6+rev2 to see if that works.

**Comment by jhermsmeier** [comment]
Okay, I just tried that with the CLI as well, and it still works for me – but there's a CLI download cache corruption bug that comes to mind: https://github.com/resin-io/resin-cli/issues/541

Might help to clear your resin-cli cache with `rm -r ~/.resin/cache/`?

**Comment by BrianAdams** [comment]
I cleared the cache and I am running it again (slow office network so it may be another 20 minutes).

I added -it to my docker command-line this time, and this is what it shows:
```
Resizing filesystem
e2fsck 1.43 (17-May-2016)
Pass 1: Checking inodes, blocks, and sizes
Pass 2: Checking directory structure
Pass 3: Checking directory connectivity
Pass 4: Checking reference counts
Pass 5: Checking group summary information
resin-data: 9653/131072 files (0.1% non-contiguous), 41915/262144 blocks
resize2fs 1.43 (17-May-2016)
Resizing the filesystem on /dev/loop3 to 456833 (4k) blocks.
The filesystem on /dev/loop3 is now 456833 (4k) blocks long.
```

**Comment by jhermsmeier** [comment]
That's looking good – the `-it` shouldn't be necessary to see that output though.

**Comment by BrianAdams** [comment]
Good news is that it works on my OSX machine and I can generate the image when i use '-it' on the command-line with my fix where I force the expansion to match the on disk size of the image.

The bad news is that when I don't run in terminal mode it fails on OSX and my linux VM (Where I cannot run in terminal mode)

```
Resizing filesystem
e2fsck 1.43 (17-May-2016)
e2fsck: need terminal for interactive repairs
```

Researching general es2sck linux guidance right now

**Comment by BrianAdams** [comment]
Added the `-p` option to the e2fsck command and it now runs in all environments

**Comment by jhermsmeier** [comment]
Awesome! I've added the `-p` flag in https://github.com/resin-io/resin-preload-image-script/pull/57 now, to avoid this happening in the future.

**Comment by jhermsmeier** [comment]
Now the only thing left is getting creative with the uncompressed layer sizes... Still not sure how we're going to get around that, as I've confirmed that there's currently no other way to get the uncompressed sizes without actually downloading them.

**Comment by BrianAdams** [comment]
I thought your approach to pull the uncompressed size using a http range query was clever. Does that not end up possible in the end?

**Comment by jhermsmeier** [comment]
> I thought your approach to pull the uncompressed size using a http range query was clever. Does that not end up possible in the end?

It took a little while figuring out how to do this in a compact fashion, but here it is: #59 – if you want to give that a go (that doesn't include the e2fsck -p fix from #57 yet though)

**Comment by BrianAdams** [comment]
#59 worked in my test!

**Comment by jhermsmeier** [comment]
I have merged #59 & #57 now, so all of these issues (apart from the resin-cli download cache) should be fixed now – so I'm going to close this for now. Feel free to reopen, if something still doesn't work as expected.
